include/revng/Pipes/ModelInvalidationRule.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include "revng/Pipeline/Container.h"
+#include "revng/Pipeline/Runner.h"
+#include "revng/Pipeline/Target.h"
+
+namespace revng::pipes {
+
+class ModelInvalidationRule {
```

Comment:
* Doc
* How is this supposed to be used? What are some examples of invalidation rules you foresee?
  I was expecting we had some form of `Kind::invalidate(TupleTreeDiff)` method.
  I see many abstractions (`InvalidationEvent`, `ModelInvalidationEvent`, `ModelInvalidationRule`) but I fail to see the endgame, which might not need all these intermediate abstractions.

Reply:
ah, no i just forgot to pass down the diff across the calls

Reply:
I still don't see how this connects to `Kind::invalidate(TupleTreeDiff)`.

Reply:
kinds must implement 

```
  virtual void getInvalidations(const pipeline::ContainerBase &Container,
                                pipeline::TargetsList &ToRemove,
                                const ModelInvalidationEvent &Event) const = 0;
```
or
```
  virtual void getInvalidations(pipeline::TargetsList &ToRemove,
                                const ModelInvalidationEvent &Event) const = 0;
```
would that work?

Reply:
Doc is still missing! Please double check before you mark as resolved.
I haven't reviewed users of `ModelInvalidationRule`, but I've this question: is this abstraction necessary?
Can't we collapse this in `Kind`? I fail to see the big picture.